### PR TITLE
Add curves

### PR DESCRIPTION
**Pull request by jpcima**
This adds a support class for curves.

- [x] parsing `<curve>` opcodes
- [x] explicit and implicit indexing
- [x] linear interpolation
- [x] spline interpolation (not in sfz, can be introduced as extension)

### CODE REVIEW

**Comment by jpcima** [comment]
Some test cases for curves (plot)

- Builtin curve 6
    - run: `./build/tests/sfizz_plot_curve 6 curves.sfz > curves.dat`
    - sfz: `/*empty*/`

- Undefined curve 7, fallback to default
    - run: `./build/tests/sfizz_plot_curve 7 curves.sfz > curves.dat`
    - sfz: `/*empty*/`

- Custom curve 7 (not indexed)
    - run: `./build/tests/sfizz_plot_curve 7 curves.sfz > curves.dat`
    - sfz: `<curve> v0=1.0 v64=0.9 v127=0.0`

- Custom curve 7, omitted side points, fallback to 0.0 and 1.0 (not indexed)
    - run: `./build/tests/sfizz_plot_curve 7 curves.sfz > curves.dat`
    - sfz: `<curve> v64=0.9`

- Custom curve 8 (indexed)
    - run: `./build/tests/sfizz_plot_curve 8 curves.sfz > curves.dat`
    - sfz: `<curve> curve_index=8 v0=1.0 v64=0.9 v127=0.0`

- Spline curve 7 (extension)
    - delete relevant `#if 0` in `Curve.cpp`
    - run: `./build/tests/sfizz_plot_curve 7 curves.sfz > curves.dat`
    - sfz: `<curve> sfizz:curve_interpolator=spline v64=0.75`

**Comment by jpcima** [comment]
Rebased on the new parser

File: src/sfizz/Curve.cpp [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+        slot = &_curves.back();
+    } else {
+        size_t index = static_cast<unsigned>(explicitIndex);
+        if (index >= _curves.size())
+            _curves.resize(index + 1);
+        _useExplicitIndexing = true;
+        slot = &_curves[index];
+    }
+
+    slot->reset(new Curve(curve));
```
Comment (paulfd) [inline src/sfizz/Curve.cpp:208]: Any reason not to just avoid the `slot` variable and call reset on `_curves.back()` or `_curves[index]`?

File: src/sfizz/Curve.cpp [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+}
+
+Curve Curve::buildPredefinedCurve(int index)
+{
+    Curve curve;
+
+    switch (index) {
+    default:
+        ASSERTFALSE;
+    case 0:
```
Comment (paulfd) [inline src/sfizz/Curve.cpp:68]: Should we break here?

File: src/sfizz/Curve.cpp [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+namespace sfz
+{
+
+static const Curve defaultCurve = Curve::buildBipolar(0.0, 1.0);
+
+Curve Curve::buildCurveFromHeader(
+    absl::Span<const Opcode> members, Interpolator itp, bool limit)
+{
+    Curve curve;
+    bool fillStatus[NumValues] = {};
```
Comment (paulfd) [inline src/sfizz/Curve.cpp:23]: I know that `std::vector<bool>` has some packing going on to avoid taking up heaps of memory. Would it be important here? Can we expect `bool[]` to have similar optimizations?

File: src/sfizz/Curve.cpp (hunk 2 above)
Comment (jpcima) [reply src/sfizz/Curve.cpp:68]: It shouldn't matter as this case does not happen.
I set the default to the linear curve.

File: src/sfizz/Curve.cpp (hunk 3 above)
Comment (jpcima) [reply src/sfizz/Curve.cpp:23]: On the plus side `vector<bool>` is compact, but access takes more instructions.
Anyway, these temporary computations are not relevant to performance.

File: src/sfizz/Curve.cpp (hunk 1 above)
Comment (jpcima) [reply src/sfizz/Curve.cpp:208]: Because this complies to RAII and avoids repeating a small code fragment in both branches.

**Comment by jpcima** [comment]
What is the bug of `lerpFill` corrected in `e58bf6f`?

**Comment by paulfd** [comment]
Any of the "Custom" test case (I missed your comment when writing them, saw it after :sweat:)
```cpp
TEST_CASE("[Curve] Custom 2")
{
    auto curve = sfz::Curve::buildCurveFromHeader({
        { "v063", "1" }
    }, sfz::Curve::Interpolator::Linear);
    REQUIRE( curve.evalCC7(0) == 0.0f );
    REQUIRE( curve.evalCC7(127) == 1.0f );
    REQUIRE( curve.evalCC7(63) == 1.0f );
    REQUIRE( curve.evalCC7(2) == Approx(0.032).margin(1e-3) );
    REQUIRE( curve.evalCC7(70) == 1.0f );
}
```
At 63 the value obtained is 0.984. I took the opportunity to use the `linearRamp` SIMD thingy although it's not active on x86, but it should be a simple index issue so we can recover the original code if you prefer. 
```
-------------------------------------------------------------------------------
[Curve] Custom 2
-------------------------------------------------------------------------------
/home/paul/source/sfizz-paulfd/tests/CurveT.cpp:119
...............................................................................

/home/paul/source/sfizz-paulfd/tests/CurveT.cpp:126: FAILED:
  REQUIRE( curve.evalCC7(63) == 1.0f )
with expansion:
  0.98437f == 1.0f
```

File: src/sfizz/Curve.cpp (hunk 2 above)
Comment (paulfd) [reply src/sfizz/Curve.cpp:68]: It warns about falling through so I'll silence it.

**Comment by jpcima** [comment]
It overwrited the index 63? oh, my bad.
The current index `iCurr` should have been checked for `fillStatus` =1, and in this case, skip it.

**Comment by jpcima** [comment]
on the `else DBG(...)` thing:

To fix this cleanly in all cases, the macro expansion should be wrapped into the form `do { ... } while (0)`.
There is a discussion: https://stackoverflow.com/questions/257418/do-while-0-what-is-it-good-for

**Comment by paulfd** [comment]
I'll let you revert it and correct it if you want. The loops that fills values are slightly tighter like this but all of this is not so critical :slightly_smiling_face: You can build it against the test suite it should cover most corner cases.

**Comment by paulfd** [comment]
Other than this I'm OK to merge!

**Comment by jpcima** [comment]
No it doesn't interpolate multi points correctly.
Check for instance `<curve> v0=1.0 v64=0.9 v100=0.9 v127=0.0`

**Comment by jpcima** [comment]
Sorry it does work correctly... it's just me who took the wrong plotfile..

**Comment by paulfd** [comment]
It's OK, your punition is another test case :+1:

**Review by paulfd — APPROVED** [review]

**Comment by jpcima** [comment]
Phew I thought I was punished to be forced to write it :smile: 
It's good to merge.
